core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,178 @@
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest list", path, exc));
+  }
+
+  private List<ManifestFile> computeManifestsToDelete(
+      Set<String> manifestListsToDelete, Map<String, ManifestFile> reachableFilesAfterExpiration) {
+    Map<String, ManifestFile> manifestCandidatesToDelete =
+        computeManifestsForManifestLists(manifestListsToDelete);
+    List<ManifestFile> manifestFilesToDelete = Lists.newArrayList();
+    for (Map.Entry<String, ManifestFile> candidateManifest :
+        manifestCandidatesToDelete.entrySet()) {
+      if (!reachableFilesAfterExpiration.containsKey(candidateManifest.getKey())) {
+        manifestFilesToDelete.add(candidateManifest.getValue());
+      }
+    }
+    return manifestFilesToDelete;
+  }
+
+  private Map<String, ManifestFile> computeManifestsForManifestLists(Set<String> manifestLists) {
+    Map<String, ManifestFile> pathToManifestFiles = Maps.newHashMap();
+    for (String manifestList : manifestLists) {
+      CloseableIterable<ManifestFile> manifestFiles = readManifestFiles(manifestList);
+      for (ManifestFile manifest : manifestFiles) {
+        pathToManifestFiles.put(manifest.path(), manifest.copy());
+      }
+    }
+    return pathToManifestFiles;
+  }
+
+  // Helper to compute files to delete
+  private Set<String> findFilesToDelete(
+      List<ManifestFile> manifestFilesToDelete, List<ManifestFile> currentManifestFiles) {
+    Set<String> filesToDelete = ConcurrentHashMap.newKeySet();
+    Tasks.foreach(manifestFilesToDelete)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn("Failed to get deleted files: this may cause orphaned data files", exc))
+        .run(
+            manifest -> {
+              // the manifest has deletes, scan it to find files to delete
+              try (ManifestReader<?> reader =
+                  ManifestFiles.open(manifest, ops.io(), ops.current().specsById())) {
+                for (ManifestEntry<?> entry : reader.entries()) {
+                  filesToDelete.add(entry.file().path().toString());
+                }
+              } catch (IOException e) {
+                throw new RuntimeIOException(e, "Failed to read manifest file: %s", manifest);
+              }
+            });
+    Tasks.foreach(currentManifestFiles)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn("Failed to get deleted files: this may cause orphaned data files", exc))
+        .run(
+            manifest -> {
+              // the manifest has deletes, scan it to find files to delete
+              try (ManifestReader<?> reader =
+                  ManifestFiles.open(manifest, ops.io(), ops.current().specsById())) {
+                for (ManifestEntry<?> entry : reader.entries()) {
```

Comment:
Since the file path is the only field that is used, this should set the projection and skip loading the other metadata into memory. This should also just iterate over the open reader rather than using entries, since none of the entry metadata is used. That will return only live data files.

Reply:
Updated , I found this utility https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/ManifestFiles.java#L93 which just reads the live entry file paths which is what we want. It does return it in string format and not charsequenceset, I could add another utility to do that but also seemed overkill?

Reply:
It should be fine. We immediately convert to String in `GenericManifestFile` anyway.